docs/_docs/04_reference/config-blocks-and-attributes.md

```diff
@@ -36,9 +36,19 @@ before or after calling Terraform.
 The `terraform` block supports the following arguments:
 
 - `source` (attribute): Specifies where to find Terraform configuration files. This parameter supports the exact same syntax as the
-  [module source](https://www.terraform.io/docs/modules/sources.html) parameter for Terraform `module` blocks, including
-  local file paths, Git URLs, and Git URLS with `ref` parameters. Terragrunt will download all the code in the repo
-  (i.e. the part before the double-slash `//`) so that relative paths work correctly between modules in that repo.
+  [module source](https://www.terraform.io/docs/modules/sources.html) parameter for Terraform `module` blocks **except
+  for the Terraform registry** (see below note), including local file paths, Git URLs, and Git URLS with `ref`
+  parameters. Terragrunt will download all the code in the repo (i.e. the part before the double-slash `//`) so that
+  relative paths work correctly between modules in that repo.
+    - NOTE: the `source` parameter can be configured to pull Terraform modules from any Terraform module registry using
+      the `tfr` protocol. The `tfr` protocol expects URLs to be provided in the format
+      `tfr://REGISTRY_HOST/MODULE_SOURCE?version=VERSION`. For example, to pull the `terraform-aws-modules/vpc/aws`
+      module from the public Terraform registry, you can use the following as the source parameter:
+      `tfr://registry.terraform.io/terraform-aws-modules/vpc/aws?version=3.3.0`. If you wish to access a private module
+      registry (e.g., [Terraform Cloud/Enterprise](https://www.terraform.io/docs/cloud/registry/index.html)), you can
+      provide the authentication to Terragrunt as an environment variable with the key `TG_TF_REGISTRY_TOKEN`. This
```

Comment:
Terraform works off a [token in a config file](https://www.terraform.io/docs/cli/config/config-file.html#credentials). Is there any way for us to reuse this existing mechanism, so you don't have to set a separate env var?

Reply:
Please read the PR description

> Similarly, I opted for using an environment variable provided API token for authenticating to private module registries. This is unconventional (not something Terraform supports), which again risks deviating from familiar APIs of Terraform, but the maintenance overhead of duplicating the logic without importing didn't seem worth it for this initial version. I think the implementation is modular enough that we can add this in later if we deem it necessary.

Reply:
Hm, alright. We should prob file a bug to track this and add a comment to the code linking to the bug.

Reply:
Done:

- https://github.com/gruntwork-io/terragrunt/issues/1771
- 55a1670